src/Avalonia.Controls/Control.cs

```diff
@@ -215,18 +280,109 @@ void IVisualBrushInitialize.EnsureInitialized()
         /// <returns>The control that receives the focus adorner.</returns>
         protected virtual IControl? GetTemplateFocusTarget() => this;
 
+        /// <summary>
+        /// Invoked as the first step of marking the control as loaded and raising the
+        /// <see cref="Loaded"/> event.
+        /// </summary>
+        internal void OnLoadedCore()
+        {
+            if (_isLoaded == false)
+            {
+                _isLoaded = true;
+                OnLoaded();
+            }
+        }
+
+        /// <summary>
+        /// Invoked as the first step of marking the control as unloaded and raising the
+        /// <see cref="Unloaded"/> event.
+        /// </summary>
+        internal void OnUnloadedCore()
+        {
+            if (_isLoaded)
+            {
+                // Remove from the loaded event queue here as a failsafe in case the control
+                // is detached before the dispatcher runs the Loaded jobs.
+                lock (_loadedQueue)
+                {
+                    _loadedQueue.Remove(this);
+                }
+
+                _isLoaded = false;
+                OnUnloaded();
+            }
+        }
+
+        /// <summary>
+        /// Invoked just before the <see cref="Loaded"/> event.
+        /// </summary>
+        protected virtual void OnLoaded()
+        {
+            var eventArgs = new RoutedEventArgs(LoadedEvent);
+            eventArgs.Source = null;
+            RaiseEvent(eventArgs);
+        }
+
+        /// <summary>
+        /// Invoked just before the <see cref="Unloaded"/> event.
+        /// </summary>
+        protected virtual void OnUnloaded()
+        {
+            var eventArgs = new RoutedEventArgs(UnloadedEvent);
+            eventArgs.Source = null;
+            RaiseEvent(eventArgs);
+        }
+
         /// <inheritdoc/>
         protected sealed override void OnAttachedToVisualTreeCore(VisualTreeAttachmentEventArgs e)
         {
             base.OnAttachedToVisualTreeCore(e);
 
             InitializeIfNeeded();
+
+            // Note: This code is not multi-threaded ready and assumes to run only on the UIThread
+            if (_isLoaded == false)
+            {
+                bool isAdded = false;
+
+                lock (_loadedQueue)
+                {
+                    isAdded = _loadedQueue.Add(this);
+                }
+
+                if (isAdded &&
+                    _isLoadedProcessingRequested == false)
+                {
+                    _isLoadedProcessingRequested = true;
+
+                    Dispatcher.UIThread.Post(static () =>
+                    {
+                        // Lock is required here since there is a possibility of the
+                        // "Collection was modified; enumeration operation may not execute." exception.
+                        // This may happen when new controls are added within the Loaded callback/event.
+                        lock (_loadedQueue)
```

Comment:
Keeping in mind that this code is always executed on the UI thread, won't it cause a deadlock if any control is attached to the visual tree inside of Loaded event handler?

Reply:
Yes, I'll document it as a feature...

This needs more thought. There seem to be no benefits with HashSet and it's going to require copying to a list for iteration. However, then you can't just call .Clear() anymore and will have to keep track of which controls have actually been processed.

Reply:
Maybe you could do something like this (pseudocode)?

```cs
if (_loadedQueue is not empty)
{
    var loadedQueue = _loadedQueue;
    _loadedQueue = new HashSet();

    // foreach here
}
```

Not sure how much overhead this would be, but maybe it would actually be possible to have 2 hash sets and swap between them, like:

```cs
if (_loadedQueue is not empty)
{
    var loadedQueue = _loadedQueue;
    _loadedQueue = _loadedQueue2;
    _loadedQueue2 = loadedQueue;

    // foreach here

    loadedQueue.Clear();
}
```

The other queue would only be visited in the next dispatcher cycle, so no conflict would occur I think?

Reply:
@jp2masa It's a simple idea to solve the concerns -- that's usually the right path. I'll consider this some more later but overall I like your idea. Thanks!

Reply:
This can also be done with a copy to a temp collection. After processing, I'll only remove the controls that are actually marked IsLoaded = true and were in the copied collection.

Since a new collection is required either way it is likely still faster to just create a new hash set and swap it out. That avoids having to iterate through the original one checking for isLoaded and then removing one by one after processing.

So I'm going to go with the swapping out for a new hash set idea. Should be fastest.

There will of course be more allocations though which really might mean the original idea of just doing this on the dispatcher for each control was comparable.

Reply:
If you use my 2nd example you only have to preallocate 2 hash sets, with the additional overhead of calling `Clear`. While you process one of the hash sets, new entries will be added to the other one, which will only be processed in the next cycle.

Reply:
Yes, good point, thanks again

Reply:
This was partially done with caveats